File: ollama/codigo-final/binary-trees/codegemma-7b/fewshot/temp_0/binary-trees.cpp
#include <cstdio>
#include <cstdlib>

struct Node {
    int data;
    Node* left;
    Node* right;
};

Node* allocate_perfect_binary_tree(int depth) {
    if (depth == 0) {
        return NULL;
    }
    Node* root = new Node;
    root->left = allocate_perfect_binary_tree(depth - 1);
    root->right = allocate_perfect_binary_tree(depth - 1);
    return root;
}

Node* allocate_long_lived_tree(int depth) {
    static Node* root = NULL;
    if (root == NULL) {
        root = allocate_perfect_binary_tree(depth);
    }
    return root;
}

int main(int argc, char* argv[]) {
    int N = atoi(argv[1]);
    int min_depth = 4;
    int max_depth = max(min_depth + 2, N);
    int stretch_depth = max_depth + 1;

    // --- Stretch Tree ---
    Node* stretch_tree = allocate_perfect_binary_tree(stretch_depth);
    printf("stretch tree of depth %d\tcheck: C\n", stretch_depth);

    // --- Long-Lived Tree ---
    Node* long_lived_tree = allocate_long_lived_tree(max_depth);
    printf("long lived tree of depth %d\tcheck: C\n", max_depth);

    // --- Bottom-Up Trees (Iterations) ---
    for (int depth = min_depth; depth <= max_depth; depth += 2) {
        int iterations = (1 << (max_depth - depth + min_depth));
        for (int i = 0; i < iterations; ++i) {
            Node* temp_tree = allocate_perfect_binary_tree(depth);
            // Walk and deallocate (e.g., `free(temp_tree)`).
        }
        printf("I\t trees of depth %d\tcheck: C\n", depth);
    }

    return 0;
}

File: ollama/codigo-final/binary-trees/codegemma-7b/fewshot/temp_0/binary-trees.py
import sys

# Define a tree node structure with left/right children.
class Node:
    def __init__(self, left=None, right=None):
        self.left = left
        self.right = right

# Function to allocate a stretch tree of given depth.
def allocate_stretch_tree(depth):
    if depth == 0:
        return None
    node = Node()
    node.left = allocate_stretch_tree(depth - 1)
    node.right = allocate_stretch_tree(depth - 1)
    return node

# Function to allocate a bottom-up tree of given depth.
def allocate_bottom_up_tree(depth):
    nodes = [None] * (2**depth - 1)
    for i in range(1, len(nodes)):
        parent_idx = (i // 2) - 1
        if parent_idx >= 0:
            nodes[parent_idx].left = nodes[i]
            nodes[parent_idx].right = nodes[i + 1] if i + 1 < len(nodes) else None
    return nodes[-1]

# Function to allocate a persistent tree of given depth.
def allocate_long_lived_tree(depth):
    return allocate_stretch_tree(depth)

# Read N from command-line arguments.
N = int(sys.argv[1])

# Set minimum and maximum depths.
min_depth = 4
max_depth = max(min_depth + 2, N)
stretch_depth = max_depth + 1

# Print header.
print("Benchmark Results:")

# Stretch Tree Case
print(f"stretch tree of depth {stretch_depth}\tcheck: allocations match reference")

# Bottom-Up Trees Case
for d in range(min_depth, max_depth + 2, 2):
    iterations = 2**(max_depth - d + min_depth)
    print(f"I\t trees of depth {d}\tcheck: allocations match reference")

# Long-Lived Tree Case
print(f"long lived tree of depth {stretch_depth}\tcheck: still exists after deallocations")
